src/mlpack/methods/svdplusplus/svdplusplus_function_impl.hpp

```diff
@@ -0,0 +1,561 @@
+      // Each subset is of size threadShareSize.
+      size_t threadId = 0;
+      #ifdef HAS_OPENMP
+        threadId = omp_get_thread_num();
+      #endif
+
+      for (size_t j = threadId * threadShareSize;
+          j < (threadId + 1) * threadShareSize && j < visitationOrder.n_elem;
+          ++j)
+      {
+        // Indices for accessing the the correct parameter columns.
+        const size_t user = data(0, visitationOrder[j]);
+        const size_t item = data(1, visitationOrder[j]) + numUsers;
+        const size_t implicitStart = numUsers + numItems;
+
+        // Prediction error for the example.
+        const double rating = data(2, visitationOrder[j]);
+        const double userBias = iterate(rank, user);
+        const double itemBias = iterate(rank, item);
+        // Iterate through each item which the user interacted with to calculate
+        // user vector.
+        arma::vec userVec(rank, arma::fill::zeros);
+        arma::sp_mat::const_iterator it = implicitData.begin_col(user);
+        arma::sp_mat::const_iterator it_end = implicitData.end_col(user);
+        size_t implicitCount = 0;
+        for (; it != it_end; it++)
+        {
+          userVec += iterate.col(implicitStart + it.row()).subvec(0, rank - 1);
+          implicitCount += 1;
+        }
+        if (implicitCount != 0)
+          userVec /= std::sqrt(implicitCount);
+        userVec += iterate.col(user).subvec(0, rank - 1);
+
+        double ratingError = rating - userBias - itemBias -
+        arma::dot(userVec, iterate.col(item).subvec(0, rank - 1));
+
+        arma::mat userVecUpdate = stepSize * (
+            lambda * iterate.col(user).subvec(0, rank - 1) -
+            ratingError * iterate.col(item).subvec(0, rank - 1));
+        arma::mat itemVecUpdate = stepSize * (
+            lambda * iterate.col(item).subvec(0, rank - 1) -
+            ratingError * userVec);
+        double userBiasUpdate = stepSize * (
+            lambda * iterate(rank, user) - ratingError);
+        double itemBiasUpdate = stepSize * (
+            lambda * iterate(rank, item) - ratingError);
+
+        // Update of item implicit vectors.
+        arma::mat itemImplicitUpdate(rank, implicitCount);
+        arma::Col<size_t> implicitItems(implicitCount);
+        it = implicitData.begin_col(user);
+        it_end = implicitData.end_col(user);
+        size_t implicitIndex = 0;
+        for (; it != it_end; it++, implicitIndex++)
+        {
+          itemImplicitUpdate.col(implicitIndex) =
+              stepSize * (lambda *
+              iterate.col(implicitStart + it.row()).subvec(0, rank - 1) -
+              ratingError / std::sqrt(implicitCount) *
+              iterate.col(item).subvec(0, rank - 1));
+          implicitItems(implicitIndex) = it.row();
+        }
+
+        // Gradient is non-zero only for the parameter columns corresponding to
+        // the example.
+        for (size_t i = 0; i < rank; ++i)
```

Comment:
I think it is wise to vectorize the loops below and use
```
#pragma omp critical
{
  iterate(i, user) -= ...
  iterate(i, item) -= ...
}
```
instead of many `omp atomic`.

Reply:
Yes but changing from `atomic` to `critical` may slow down the program execution.

Reply:
Perhaps you are right. I don't know exactly how `atomic` is implemented internally, but it is possible `atomic` doesn't deal with mutexes.